File: packages/decoder/src/mlp_up.ts
import { checkNotNull } from './check_not_null.js';

export interface MlpUpInit {
  readonly inputSize: number;
  readonly outputSize: number;
  readonly numLayers: number;
}

export class MlpUp {
  static wasmModule: WebAssembly.Module;

  static async create(init: MlpUpInit): Promise<MlpUp> {
    return new MlpUp(init, await WebAssembly.instantiate(this.wasmModule));
  }

  readonly normWeight: Uint8Array;
  readonly gateWeight: Uint8Array;
  readonly upWeight: Uint8Array;
  readonly inputVector: Float32Array;
  readonly outputVector: Float32Array;

  private readonly self: number;

  constructor(
    { inputSize, outputSize, numLayers }: MlpUpInit,
    private readonly wasmInstance: WebAssembly.Instance,
  ) {
    const {
      memory,
      create,
      getNormWeight,
      getGateWeight,
      getUpWeight,
      getInputVector,
      getOutputVector,
    } = wasmInstance.exports as WasmExports;

    this.self = create(inputSize, outputSize, numLayers);

    this.normWeight = new Uint8Array(
      memory.buffer,
      checkNotNull(getNormWeight(this.self)),
      numLayers * inputSize * 4,
    );

    this.gateWeight = new Uint8Array(
      memory.buffer,
      checkNotNull(getGateWeight(this.self)),
      numLayers * outputSize * inputSize * 4,
    );

    this.upWeight = new Uint8Array(
      memory.buffer,
      checkNotNull(getUpWeight(this.self)),
      numLayers * outputSize * inputSize * 4,
    );

    this.inputVector = new Float32Array(
      memory.buffer,
      checkNotNull(getInputVector(this.self)),
      inputSize,
    );

    this.outputVector = new Float32Array(
      memory.buffer,
      checkNotNull(getOutputVector(this.self)),
      outputSize,
    );
  }

  forward({ layer }: Readonly<{ layer: number }>): void {
    (this.wasmInstance.exports as WasmExports).forward(this.self, layer);
  }
}

interface WasmExports extends WebAssembly.Exports {
  readonly memory: WebAssembly.Memory;

  create(inputSize: number, outputSize: number, numLayers: number): number;
  getNormWeight(self: number): number;
  getGateWeight(self: number): number;
  getUpWeight(self: number): number;
  getInputVector(self: number): number;
  getOutputVector(self: number): number;
  forward(self: number, layer: number): void;
}

File: src/mlp_up.zig
const std = @import("std");

const addition = @import("simd/addition.zig");
const hadamard_product = @import("simd/hadamard_product.zig");
const matrix_vector_product = @import("simd/matrix_vector_product.zig");
const rms_norm = @import("simd/rms_norm.zig");
const swish = @import("simd/swish.zig");

const Self = @This();

norm_weight: []f32,
gate_weight: []f32,
up_weight: []f32,
input_vector: []f32,
output_vector: []f32,

// private
norm_vector: []f32,
gate_vector: []f32,

export fn create(input_size: usize, output_size: usize, num_layers: usize) ?*Self {
    const allocator = std.heap.page_allocator;
    const self = allocator.create(Self) catch return null;

    self.* = .{
        .norm_weight = allocator.alloc(f32, num_layers * input_size) catch return null,
        .gate_weight = allocator.alloc(f32, num_layers * output_size * input_size) catch return null,
        .up_weight = allocator.alloc(f32, num_layers * output_size * input_size) catch return null,
        .input_vector = allocator.alloc(f32, input_size) catch return null,
        .output_vector = allocator.alloc(f32, output_size) catch return null,
        .norm_vector = allocator.alloc(f32, input_size) catch return null,
        .gate_vector = allocator.alloc(f32, output_size) catch return null,
    };

    return self;
}

export fn getNormWeight(self: *const Self) [*]f32 {
    return self.norm_weight.ptr;
}

export fn getGateWeight(self: *const Self) [*]f32 {
    return self.gate_weight.ptr;
}

export fn getUpWeight(self: *const Self) [*]f32 {
    return self.up_weight.ptr;
}

export fn getInputVector(self: *const Self) [*]f32 {
    return self.input_vector.ptr;
}

export fn getOutputVector(self: *const Self) [*]f32 {
    return self.output_vector.ptr;
}

export fn forward(self: *const Self, layer: usize) void {
    const norm_weight_size = self.input_vector.len;

    rms_norm.compute(
        self.norm_weight[layer * norm_weight_size ..][0..norm_weight_size],
        self.input_vector,
        self.norm_vector,
    );

    const up_weight_size = self.output_vector.len * self.input_vector.len;

    matrix_vector_product.compute(
        self.gate_weight[layer * up_weight_size ..][0..up_weight_size],
        self.norm_vector,
        self.gate_vector,
    );

    swish.compute(self.gate_vector, self.gate_vector);

    matrix_vector_product.compute(
        self.up_weight[layer * up_weight_size ..][0..up_weight_size],
        self.norm_vector,
        self.output_vector,
    );

    hadamard_product.compute(self.output_vector, self.gate_vector, self.output_vector);
}
